server/src/main/scala/com/cloudera/livy/server/interactive/InteractiveSession.scala

```diff
@@ -369,6 +369,18 @@ class InteractiveSession(
     info(msg)
     sessionLog = IndexedSeq(msg)
   } else {
+    val uriFuture = future { client.get.getServerUri.get() }
+
+    uriFuture onSuccess { case url =>
+      rscDriverUri = Option(url)
+      sessionSaveLock.synchronized {
+        sessionStore.save(RECOVERY_SESSION_TYPE, recoveryMetadata)
+      }
+    }
+    uriFuture onFailure { case e =>
+      throw new IllegalStateException("Fail to get rsc uri", e)
```

Comment:
Instead of throwing an exception here, I think we can print a warning. If the RSC Driver is indeed dead, PingJob will fail and its listener will stop the session properly.